RFC: Reduce API fragmentation across platforms (#496)
react-native-community/discussions-and-proposals: proposals/0000-reduce-fragmentation.md

This adds redundancy to the React Native API, as the proposal is to make additive changes and avoid requiring removal of existing APIs in the short term.

We have to be very intentional about how much compatibility with open web standards we can promise. Some cross-platform limitations are rooted in significant differences in the host platforms APIs (e.g., accessibility) and the lack of 99% compat may frustrate users.

Aiming for a degree of React DOM compatibility shifts the burden of missing features onto React Native. Developers may not be used to an API that is complete on web but partially implemented on native platforms.

-----

# Environment APIs

## `window`

### Events

* [ ] Receives all W3C events dispatched to components (capture and bubble phase). Resolves proposal [#249](https://github.com/react-native-community/discussions-and-proposals/issues/249).
* [ ] Dispatches [`resize`](https://developer.mozilla.org/en-US/docs/Web/API/Window/resize_event).

### Properties

> * [ ] [`devicePixelRatio`](https://developer.mozilla.org/en-US/docs/Web/API/Window/devicePixelRatio)

Reviewer: This is where it's going to start to get hard to align web and native.  We already hit issues with the existing RN PixelRatio API being too simplistic for native scenarios (especially on desktop scenarios).  In particular, with a desktop application you can have multiple windows running within the same RN instance. There cannot be a concept of a global window.  These windows may be on different monitors, each with a different decicePixelRatio.  This is complexity that as far as I know web does not have to deal with since each window has its own JS instance, so the W3C APIs are not sufficient, and in fact promote incorrect over simplification of the native platform.

* [ ] [`navigator`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator)
  * [ ] [`clipboard`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/clipboard)
  * [ ] [`languages`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/languages)
  * [ ] [`permissions`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/permissions)
  * [ ] [`vibrate()`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/vibrate)

> * [`performance`](https://developer.mozilla.org/en-US/docs/Web/API/performance_property)

Reviewer 2: Would be great if these were installed directly on the `window` object from native so we can reduce the amount of runtime JS required to create the expected runtime. Tons of shims results in more complicated bundling and larger bundles to debug.

Author: Isn't that what's proposed here?

"Tons of shims"... what shims are you referencing?

* [ ] [`p`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/p).
* [ ] [`pre`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/pre).

## Inline text

Inline text semantics define the meaning, structure, or style of a word, line, or any arbitrary piece of text.

* [ ] [`a`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a).
* [ ] [`b`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/b).
* [ ] [`bdi`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/bdi).
* [ ] [`bdo`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/bdo).
* [ ] [`code`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/code).
* [ ] [`em`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/em).
* [ ] [`span`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/span) is equivalent to `Text` with `display:inline` layout.
* [ ] [`strong`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/strong).
* [ ] [`sub`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/sub).
* [ ] [`sup`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/sup).

## Media

> * [ ] [`img`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img) is similar to `Image` without advanced loader configuration.

Reviewer 2: Implementing `picture` with fallbacks to different supported sources would be great, especially since there's a high probability that a custom runtime may not support a the same formats across platforms. e.g. `.heic` on iOS with fallback to `jpg` on web.

Author: Maybe at some point. But `picture` has very little use inside Meta, and isn't commonly used outside Meta from what I can see.

* [ ] [`button`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button).
* [ ] [`input`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) is equivalent to `TextInput` with `inputMode` corresponding to `type`. Expanding input type support to number, datetime, color, etc., resolves proposal [#510](https://github.com/react-native-community/discussions-and-proposals/issues/510).
* [ ] [`label`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label).
* [ ] [`optgroup`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/optgroup).
* [ ] [`option`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option).
* [ ] [`progress`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/progress).
* [ ] [`select`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select).
* [ ] [`textarea`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea) is equivalent to `TextInput` with `multiline={true}`.

> 

Reviewer 2: `<table>` and all associated elements seem important too given how prominent they are in markdown and MDX on web.

Author: Do OSS stakeholders really want tables and have an idea of how to implement that in React Native?

To avoid breaking changes in React Native and React DOM, it may be preferrable to create new event prop names that receive events as specified by web standards, e.g., `on-click`, `on-pointerup`.

Resolves proposal [#492](https://github.com/react-native-community/discussions-and-proposals/issues/492).

- [ ] `on-auxclick` is a `PointerEvent` for auxillary clicks.
- [ ] `on-blur`.
- [ ] `on-click` is a `PointerEvent` for primary clicks.
- [ ] `on-contextmenu` is a `PointerEvent` for context clicks / long press.
- [ ] `on-copy`.
- [ ] `on-cut`.
- [ ] `on-focus`.
- [ ] `on-gotpointercapture`.
- [ ] `on-lostpointercapture`.
- [ ] `on-paste`.
- [ ] `on-pointercancel`.
- [ ] `on-pointerdown`.
- [ ] `on-pointerenter`.
- [ ] `on-pointerleave`.
- [ ] `on-pointermove`.
- [ ] `on-pointerout`.
- [ ] `on-pointerover`.
- [ ] `on-pointerup`.
- [ ] `on-keydown`.
- [ ] `on-keyup`.

> ## `<a>` props

Reviewer 2: Perhaps the most critical difference between web and native is how you can use strings as children in any component (used heavily in conjunction with the anchor element). Would be good to know where this project stands on adding that functionality to native.

Author: Auto-wrapping strings might be an option. More complicated is breaking the expectation that you can set text styles on random elements far up the tree and expect them to be inherited by elements all the way down. I suspect that isn't going to be on the cards.

Reviewer 3: > More complicated is breaking the expectation that you can set text styles on random elements far up the tree and expect them to be inherited by elements all the way down.

I think inheriting text styles like that is unnecessary. Not crashing your entire app from `{"" && <div />}` already makes a big difference.

* [ ] [`backgroundImage`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-image). Add support for setting background images via `url()`. Stretch: support CSS gradients.
* [ ] [`backgroundOrigin`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-origin).
* [ ] [`backgroundPosition`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-position).
* [ ] [`backgroundRepeat`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-repeat).
* [ ] [`backgroundSize`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-size).
* [ ] [`boxSizing`](https://developer.mozilla.org/en-US/docs/Web/CSS/box-sizing).
* [ ] [`boxShadow`](https://developer.mozilla.org/en-US/docs/Web/CSS/box-shadow). Add native support for CSS box shadows to replace buggy, iOS-specific `shadow*` styles. Resolves issue (#26110)[https://github.com/facebook/react-native/pull/26110] and [multiple other issues](https://github.com/facebook/react-native/search?q=ios+shadow&type=issues).
* [ ] [`caretColor`](https://developer.mozilla.org/en-US/docs/Web/CSS/caret-color).
* [ ] [`clipPath`](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path).

> * [ ] [`display`](https://developer.mozilla.org/en-US/docs/Web/CSS/display) values of `block`, `contents` (resolves proposal [#348](https://github.com/react-native-community/discussions-and-proposals/issues/348)), `inline`, `inline-block`, `inline-flex`.

Reviewer 4: If the goal is "the best bits of the web" then I feel like `display: grid` really ought to be supported too.

Reviewer 5: The layout engine is [Yoga](https://github.com/facebook/yoga), and it doesn't support grid (because grid is fairly complicated to build on top of the native platforms).

I feel like grid should be a part of _"Make RN more like Web V2"_ to keep from holding anything else up.

- https://github.com/facebook/yoga/issues/47 (original "support grid" issue from 2015)
- https://github.com/facebook/yoga/issues/867 (2019 updated issue)

Reviewer 4: @corysimmons It doesn't. But then Yoga also doesn't support `block`, `inline` or `float` layout, and they're all included in this proposal. Even Yoga's Flexbox implementation isn't web compatible and will need a rework if the aim is to support web layouts without modification.

Grid is complicated, but it's not really much more complicated than flexbox. I link to an [existing native code implementation](https://github.com/DioxusLabs/taffy) that could be either be used alongside Yoga or ported into Yoga in the issue you mention (https://github.com/facebook/yoga/issues/867) (disclaimer: I wrote said implementation - but this is not me just shilling my own library - I wrote it in large part *because* I want to see it in React Native)

Is Grid just considered less essential? It's one of the most highly requested feature by React Native end users (me being one them!). Surely it at least ought to be considered higher priority than `float`, no?

Reviewer 6: Yeah, offline I had advocated for "grid" before "block" or "inline", and I do think it would be likely to come first (though I would also believe there is more web code using the latter if porting existing web code is the main draw).

It is exciting that Taffy has an implementation of Grid as well, though I want to set expectations that we are not planning to sequence work to port grid in the short term. I don't think we have solved safely innovating on Yoga yet, nor do I think we have really solved problems which would let us smoothly transition RN to any alternative layout engines. Both are at the scale of hundreds of thousand of surfaces, and billions of users, and even small changes in behavior of performance characteristics can effectively take out major portions of the industry.

To create better clarity there is a plan I am going to share out for concrete Yoga + RN work we are taking on over the next couple of months, as a more concrete articulation of the vision and accompanying discussion shared late last year. The majority of it is around shaping Yoga/RN to start letting us make conformance changes more confidently, along with getting a new OSS release of Yoga out.

Reviewer 4: @NickGerleman I guess I'm pretty skeptical of what appears to be the proposed approach. Which as I understand it is something like the following:

> Incrementally evolve the existing algorithm towards standards compliance using fine-grained feature flags to alter behaviour in each place where the standards-compliant behaviour differs from the legacy behaviour while keeping a shared implementation for parts where the behaviour is the same.

It seems to me that by taking such an approach you end up with what is essentially two different algorithms sharing an implementation and that this is good nobody:

- **It's not good for users of React Native who's apps are in maintenance mode and need strict backwards compatibility**, because it means that almost any change to Yoga, even if it aimed at the new "web compatible" version of the algorithm, is likely to touch codepaths that their app uses. This greatly increases the chance of a regression / backwards incompatible change being introduced.

- **It's not good for users of React Native who want web-compatible layout** because it means development will be much slower (see point 3) Which means that either such an implementation will not be available at all for some time to come (years?), or that there will ongoing breaking changes once they adopt the algorithm. Which may well not tolerable even for people who willing to do a one-time pass over their code to switch to a new algorithm.

- **It's not good for maintainers of / contributors to Yoga** as it will be much harder to make sense of each algorithm's flow when the implementations of each are intertwined with each other. And also because it meants that every change has to be checked for compatibility with the legacy algorithm (even if it is only intended to affect the new one), for which Yoga does not have a good test suite and testing is slow. This also effectively all-but-excludes OSS contributors from contributing algorithm changes to Yoga as they will not have access to the testing infrastructure that Meta employees are using the to test for compatibility with legacy apps.

I also think you might be underestimating the amount of changes that would be required to make Yoga's current implementation standards compliant. Based on my experience making similar changes to Taffy (which is derived from an earlier version of Yoga, and even after the changes made so far is still not entirely standards compliant), I would not be at all surprised if a shared implementation of RN's current algorthm and a standards-compliant flexbox algorithm ended up with more code feature flagged than shared. I would expect the tweaks to be small, but large in number and widespread throughout the algorithm.

---

My counter-proposal would be something like the following:

1. Adapt the current implementation such that rather than recursively calling itself when it needs to size or layout a child node, it calls into a `layout_child` function that it takes as a function pointer parameter (or some such similar arrangement that allows the caller to provide a function to call to size child nodes). This will allow either Yoga itself or the embedder of Yoga to take back control flow between each node in the hierarchy and enables the possibility of switching between different algorithms/implementatons for each node.

2. Freeze the current algorithm as it is. Aside from things like minor changes for compatibility with new compiler or similar, this code doesn't get touched from this point onwards.

3. Implement a new standards-compliant algorithm. This could either be a copy of the current algorithm adapted to be more standards compliant (but free from the burden of backwards compatibility). Or it could be a port of an existing implementation that is already (more) standards compliant. Or it could be directly calling into an existing implementation that is already (more) standards compliant.

4. In terms of releasing the new algorithm to RN users: it should be done in an opt-in manner on a per-node basis. I would suggest something like a new `Div` component (or `Flex`/`Row`/`Column` components), leaving `View` to continue using the legacy algorithm indefinitely, but there are a few different ways that you could handle this. I also think some effort should be put into tightening up the standards-compliance of the new implementation before release (perhaps porting the WPT test suite) so as to largely avoid the need for further breaking changes (although it could potentially be released as a preview version first, similarly to how to Hermes was introduced).

Step 1 would need to be done carefully in order to avoid breaking backwards compatibility, but no more carefully than *every single change* would need to be made if you go down the shared implementation route. But once you've made that changes then *most* algorithm changes to the new algorithm wouldn't need nearly as much scrutiny as legacy users simply wouldn't use that codepath at all. And as far as I can see, such a change would need to be made anyway if you want to support mutliple layout algorithms (even ones as simple as `display: block`).

Reviewer 6: > It seems to me that by taking such an approach you end up with what is essentially two different algorithms sharing an implementation and that this is good nobody:

We end up with different branches in Yoga, for each case we find a bug. And a way in Yoga to specify whether we want conformance or compatibility, to let users of Yoga control the breaks. StrictLayout is the way we expose this to React Native, and provides immediate value to cross-users with web, but the plan is to encourage migration via warnings of relying on inconsistent behavior. This does add some debt to Yoga, when we discover a behavior is relied upon. Though we have a framework for deciding what changes are encountered rarely enough to make the break.

---

> I also think you might be underestimating the amount of changes that would be required to make Yoga's current implementation standards compliant.

Followed by 

> Implement a new standards-compliant algorithm. This could either be a copy of the current algorithm adapted to be more standards compliant (but free from the burden of backwards compatibility).

We do though have a queue already of conformance issues we know about, and so far the implementation has not been the tricky bit. Because we have the existing queue, I am seeing WPT exploration as a lower priority, until we start landing the fixes. Though I would consider them to be a requirement for adding new features/algorithms.

---

> Adapt the current implementation such that rather than recursively calling itself when it needs to size or layout a child node, it calls into a layout_child

The design behind `LayoutableShadowNode` has this sort of scenario in mind in Fabric. This is the sort of thing I think we could do if we wanted escape hatches to different layout systems per-node. Historically I think this has come up with if we wanted RN to have a lower level concept like `HStack` or `VStack`, since we have heard from other teams this can be implemented more performantly as its own primitive instead of directly on top of flexbox.

---

> It's not good for users of React Native who's apps are in maintenance mode and need strict backwards compatibility, because it means that almost any change to Yoga, even if it aimed at the new "web compatible" version of the algorithm, is likely to touch codepaths that their app uses.

> Freeze the current algorithm as it is.

I have been exploring the strategy of freezing the algorithm, but am not sure we should follow it, as we will need to maintain both new and old Yoga then. As they diverge, this becomes harder, which is why I have so far been more interested in instead allowing Yoga to switch its own algorithms or branch for bugfixes.

So we do instead continue to change Yoga, with an intentionality to fixes. Though specific users of Yoga may continue to use older versions if they are hesitant to use a layout engine which has any active work being done.

---
 
> In terms of releasing the new algorithm to RN users: it should be done in an opt-in manner on a per-node basis. 

Is the exact strategy we are taking with StrictLayout. This is just controlled by a per-Node Yoga config, instead of a new layout engine.  

---

In other threads I have seen your advocacy of replacing Yoga with Taffy and adding a per-node flag to switch to the other engine, and I think maybe this is some of the motivation here? I have seen Taffy as a healthy project, and have been impressed by the work happening, but a lot of this has been coming with a suggestion that Taffy is more conformant, feature-rich, and conformant as Yoga. Apart from some gaps I think you are aware of (e.g. RTL support), I think we would need a lot more evidence, to make these sorts of assertions.

Reviewer 4: > I have been exploring the strategy of freezing the algorithm, but am not sure we should follow it, as we will need to maintain both new and old Yoga then. As they diverge, this becomes harder, which is why I have so far been more interested in instead allowing Yoga to switch its own algorithms or branch for bugfixes.

As I see it, the old algorithm shouldn't require much maintenance. Only occasional minor fixes to ensure it continues to compile on newer compilers (which given C++'s generally good backwards compatibility story should be minimal). There should be no need to make any algorithm changes to the frozen algorithm. If users want changes or fixes then they can migrate to the new algorithm. The algorithms diverging over time therefore doesn't create any extra work under a model where it is frozen and ongoing changes happen in a copy, as you're only making meaningful changes to the new algorithm.

However, if the old algorithm and the new algorithm share code then I think *that* will make the code harder and harder to maintain as the algorithms diverge. Because:

1. You end up with branches in the code all over the place, making it hard to read the code and mentally walk through it's execution

2. Any changes made to shared codepaths have to be validated for correctness against two different standards. One of which (the current behaviour) we do not a test suite for.

---

Regarding the latter point. You say that:

> we have a framework for deciding what changes are encountered rarely enough to then only include new behavior.

However my understanding is that this process amounts to something like "make the change in a branch of the meta monorepo and then run CI across everything affected". And while this is a useful tool to have it is:

1. Not available to potential contributors outside of Meta

2. A relatively slow process to run.

It does not seem like the sort of tool like that could be run iteratively after making a small code change to quickly gain feedback on whether that change has fixed and/or regressed correct behaviour. And I see that as a big problem:

Suppose I (or some other OSS contributor) contributes a PR that updates Yoga to be more standards conformant. They make the changes and include generated tests to show that the changes are correct according to web standards. They then submit their PR but it gets rejected because it breaks the legacy algorithm. Their PR is now effectively blocked. And what's more, that contributor is not in a position to fix it because they are unable to verify if their changes are correct other than via a process of pushing changes, pinging a Meta employee to test their changes and report back. You'd likely end up with iteration cycles measured in days if not weeks. That's not a practical way to develop, and I certainly can't see myself contributing in such an environment.

If you split the algorithms at a high level then you still have this same problem for the frozen algorithm. But as changes to that algorithm would be both rare and minor this is much less of an issue. Anyone is empowered to make major changes to the new version of the algorithm, and has the tools available to verify the correctness of those changes.

IMO this would have a huge impact on development velocity.

---

Regarding how changes are exposed in React Native: I think we are largely in agreement.

---

Regarding Taffy:

- It's true that I am motivated to have it used. I spent a long time working on implementing CSS Grid in Taffy and improving the performance and spec conformance of it's Flexbox implementation, and I would love to see that work being made use of. Especially in React Native, as I am a user of RN myself, and would therefore get to benefit personally from improvements to it's layout capabilities.
- *However*, I also genuinely believe that a split algorithm approach is best regardless of whether Taffy is used. Taffy seems like the quickest way to get access to new algorithms (CSS Grid and standards conformant Flexbox) in Yoga, as you can call a function from a library rather than having to write and validate the entire algorithm. But if using Taffy is considered off the table (perhaps because it's in Rust, or because the code is not owned by Meta), I would still advocate for writing the new algorithm separately from the existing one for the reasons explained above.

Regarding Taffy's performance:

- We have benchmarks showing that we're comparable in performance to Yoga (see https://github.com/DioxusLabs/taffy#benchmarks-vs-yoga). We've actually improved performance a little since those were measured. In general, Taffy tends to faster in "deep" benchmarks (few nodes per container but more levels of nodes) and Yoga faster in "wide" benchmarks (many nodes per container but fewer levels), but there isn't much in it either way.

Regarding Taffy's spec conformance:

- Taffy is a fork of [Stretch](https://github.com/vislyhq/stretch) which was created by Emil Sjölander who was also a key contributor to Yoga between 2016 and 2018. He has [claimed](https://news.ycombinator.com/item?id=19781441) that Stretch was more web compatible than Yoga and was in a position to know.
- Certainly it is the case that Taffy, via a combination of the changes made in Stretch and further changes to Taffy itself, no longer suffers from the same deviations from the standard that Yoga is known to have around min/max sizes/constraints and aspect ratio (documented  [here](https://github.com/facebook/yoga/blob/b17f08ed4915f24b794a7f2ffdc5cd0acfaf88e1/yoga/Yoga.cpp#L2663) and [here](https://github.com/facebook/yoga/blob/b17f08ed4915f24b794a7f2ffdc5cd0acfaf88e1/yoga/Yoga.cpp#L781).
- Taffy passes a superset of Yoga's test suite. This includes all tests from both Yoga's `main` and the imported tests from FlexLayout (https://github.com/facebook/yoga/pull/1194) which amounts to 264 tests + an additional 177 tests of our own. I'm not sure how Yoga would do on those tests, but that might be an interesting experiment to do.
- That said, Taffy certainly isn't 100% spec compliant either. There are 3 known issues I am aware of with Taffy:
   - No RTL support
   - No support for `overflow` (although it's affect on layout is minor)
   - Taffy incorrectly resolves percentages against min-heights where it should refuse to.
   
Personally I think Taffy would represent an easier starting point for achieving spec compliant Flexbox (and better than starting from scratch for achieving CSS Grid!). But that may be influenced my better familiarity with Taffy, and also that my Rust is stronger than my C++.

---

Regarding "supporting multiple layout engines":

>  I am not wanting to put us in a position where we support multiple layout engines long term.

That seems reasonable. And maybe that's a good reason to have in-tree implementations of the new web compatible algorithms. But from the way you are talking about things upthread, development of such algorithms is not considered a priority and there are no immediate plans to carry out development of that kind. What is the point of an RFC like this if it isn't actually going to be implemented?

It seems to me that either depending on another library or porting it's algorithms into Yoga would be a good way to get a headstart on such development and make it practical to implement these things in a reasonable timeframe. I see no reason, for example, why React Native shouldn't have both Standards Compliant Flexbox and CSS Grid available by the end of the calendar year (in preview at least).

Author: @nicoburns I'm sympathetic to your perspective and would like React Native to be able to benefit from your experience, expertise, and work. I like the idea of getting to this "end state" API sooner rather than later, if only to (in)validate its worth. And we have limited engineering time at Meta. Looking through the Taffy issues, etc., I wonder what your view on how we could support flow layout (display block, inline, inline-block...even `float`) on native? As much as people want CSS Grid too, at the moment we can't even render a `<div>` with the same default layout as a browser would, but that's what a lot of existing React DOM code relies on.

Reviewer 4: @necolas `display: block` should be quite straightforward (which would at least solve the simplest cases of `<div>`s using the default (block) layout)

---

However, `display: inline`, `display: inline-block` and `float` would be much more complicated. This is because these layout modes interact closely with text layout. For this reason, they couldn't really be implemented in either Yoga or Taffy.

In the Rust GUI world the goto library for text layout is [cosmic-text](https://github.com/pop-os/cosmic-text), but cosmic-text assumes that you will be do low-level rendering and handling things like text-selection using raw input events (i.e. more of a Flutter approach than a React Native approach to rendering).

I believe React Native uses the platform's native text-rendering capabilities. So I think it would be limited to whatever is supported by those. I'm not familiar with iOS/Android native text view support. So I'm not sure how much is even possible. I know that React Native supports images in Text components, so some kind of content-inline-with-text must be possible. But I also know that View inside Text is not supported / is buggy. And it's potentially a lot more complex when you consider things like the fact that an `inline-block` box might itself contain text, and that text selection should work seamless across the boundary!

So I think that most of an implementation of `inline/inline-block/float` would need to be in the platform-specific code rather than a layout library like Yoga/Taffy. Perhaps support for laying subtrees could be improved, although I suspect this could be made to work with Yoga's existing measure_func support.

---

If the goal is to render "typical" web content as-is, there are a couple of other features which I think merit consideration:

- `position: static`: i.e. the ability to position `position: absolute` relative to a non-direct-parent ancestor node. Neither Taffy nor Yoga implement this currently (Yoga has support for specifying this style, but as far as I can tell it doesn't actually use it to alter the layout!). Quite a lot of web layouts use this IME (mine do anyway). This shouldn't be too hard to add. The main issues would be threading a context down with the correct offsets, and perhaps adding a way to record which ancestor the node ends up being relative too (although this may not be necessary). There may also be non-layout related complexities as IIRC on web it's possible for a `position: absolute` node to "escape" a scroll view and not scroll despite being a child of that scroll view.

- Table layout. I would definitely consider this in scope for Taffy. However, it's pretty complex (at least non-fixed table layout) and I suspect it would probably be at least a 6 month project for someone to implement it.

---

Finally, it seems worth mentioning that there is this library which implements some of this on top of the existing React Native, but it's unsurprisingly quite limited https://meliorence.github.io/react-native-render-html/docs/reinvent-the-wheel

Author: Good to know. I think supporting `display:block` in some form is the most significant part of unblocking initial compatibility for element layout. Are you saying you think that's relatively straight-forward to add to Yoga too?

I think RN can already approximate inline text layout, static positioning, and even flexbox well enough to unblock some next steps related to DOM compat...if we set some restrictions in place for now (e.g., no `inline` support for block elements). My interest in primarily in getting `div` rendering as you'd expect on the web; not in changing the default layout used by `View` in existing surfaces.

Reviewer 4: > Are you saying you think that's relatively straight-forward to add to Yoga too?

Yes, I think so. I think you'd need to implement changes similar to https://github.com/DioxusLabs/taffy/pull/246 which factors out child sizing so that it isn't assumed that a child is `display: flex`.

The block layout algorithm itself is pretty simple (https://www.w3.org/TR/CSS22/visuren.html#block-formatting), although it should be noted that there is a decent amount of complexity introduced by margin collapsing (https://www.w3.org/TR/CSS22/box.html#collapsing-margins), in particular because margins can collapse between parents and children rather than just between siblings. Which I think means that the block algorithm may need to recurse arbitrarily far down the tree whereas the flexbox algorithm only need to look at the styles of direct children.

> My interest in primarily in getting div rendering as you'd expect on the web; not in changing the default layout used by View in existing surfaces.

If you're happy with how Text works and don't need `display: inline-block` then that seems achievable to me :)

Author: Well there's always room for improvement but I'm hopeful that this would be enough to motivate further changes in the future! Thanks for taking the time to help me learn a bit more about all this

Reviewer 4: @necolas Taffy has just landed (https://github.com/DioxusLabs/taffy/pull/474) block layout support. The implementation (https://github.com/DioxusLabs/taffy/blob/main/src/compute/block.rs) was actually suprisingly straightforward (sitting at ~650 LoC, with a couple of relatively minor tweaks to the interface we use to allow parents to measure the size of their children also being necessary in order to accomodate margin collapsing). And we have just over 200 (yoga-compatible) test fixtures to go with it.

However, I have another question for you around supporting `<div>`s with default styles: would you consider `box-sizing: content-box` support necesary? Because Yoga and Taffy both only support `box-sizing: border-box` currently. I figure that most of the web is resetting `box-sizing` to `border-box` but technically it isn't the default.

Author: I think `context-box` is another "improvement" rather than blocker. For context, the way I've implemented this API is to create a new library that exports "strict" DOM components. A cross-platform component looks like this:

```js
import { html } from 'react-strict-dom';
import { stylex } from '@stylexjs/stylex';

const styles = stylex.create({
  root: {}
});

function MyComponent(props) {
  return (
    <html.div style={styles.root} {...props} />
  );
}
```

The `html` components are as designed in this spec, where `style` is a special prop for all platforms - it's more like the React Native one that then React DOM one. The `html` components on web come with a built-in style reset that normalizes the layout across web and native, including setting `box-sizing:border-box`. There's no way around that to get this off the ground, so we'll probably be "stuck" with `border-box` as the default long-term (which I expect most web products will be ok with, including if they migrate from `react-dom`). But eventually there may be benefits to including support for `content-box` too.

At least, that's my thinking right now.

Author: Following up on my previous comment. Internally we've had engineers surface rendering issues that were traced back to elements being styled with `context-box` as the assumed `box-sizing`, and switching those elements to `border-box` was an unexpected event that required style modifications to avoid regressions. So I think in the short-term, it may be an acceptable cost for early adopters rendering existing React DOM UI on native to have to experience regressions like this and do manual adjustments. But if you're trying to "migrate" larger amounts of existing React DOM code to render on native, the native renderer will eventually need the default <div>/<span> layouts to be accurate. I think this is especially valuable for UI because you often only see visual regressions once they've already broken something in production, unless your application has extensive screenshot testing (which most don't/can't).

  * `aria-modal` is the same as `accessibilityViewIsModal`.
  * `aria-selected` is the same as `accessibilityState.selected`.
  * `aria-valuemax` is the same as `accessibilityValue.max`.
  * `aria-valuemin` is the same as `accessibilityValue.min`.
  * `aria-valuenow` is the same as `accessibilityValue.now`.
  * `aria-valuetext` is the same as `accessibilityValue.text`.
  * `role` is the same as `accessibilityRole` (with support for web values).
  * See prior work: [RFC Accessibility APIs](https://github.com/react-native-community/discussions-and-proposals/pull/410/files).
* Add `dir` prop to configure element's layout / writing direction.
* Make `id` the same as `nativeID`.
* Make `onAuxClick` the same as React DOM `onAuxClick` PointerEvent (can't be shimmed).
* Make `onClick` the same as React DOM `onClick` as PointerEvent (can't be shimmed).
* Make `onPointer*` props that same as React DOM PointerEvent props (can't be shimmed).
  * The Android and iOS implementation is currently being developed by Meta.
* Make `tabIndex` the same as React Native `focusable` prop.
  * Add support for `0` and `-1` values only.

Example:

```jsx
<View

>   aria-hidden={false}

Reviewer 7: I wonder if there are any gotchas in the infra that need to be fixed to support prop names with dashes...

  off: 'off',
  'one-time-code': 'sms-otp',
  'postal-code': 'postal-code',
  sex: 'gender',
  'street-address': 'street-address',
  tel: 'tel',
  'tel-country-code': 'tel-country-code',
  'tel-national': 'tel-national',
  username: 'username'
};

// https://reactnative.dev/docs/textinput#textcontenttype-ios
const autoCompleteIOS = {
  'address-line1': 'streetAddressLine1',
  'address-line2': 'streetAddressLine2',
  'cc-number': 'creditCardNumber',
  'current-password': 'password',
  country: 'countryName',
  email: 'emailAddress',
  name: 'name',
  'additional-name': 'middleName',
  'family-name': 'familyName',
  'given-name': 'givenName',
  nickname: 'nickname',
  'honorific-prefix': 'name-prefix',
  'honorific-suffix': 'name-suffix',
  'new-password': 'newPassword',
  off: 'none',
  'one-time-code': 'oneTimeCode',
  organization: 'organizationName',
  'organization-title': 'jobTitle',
  'postal-code': 'postalCode',
  'street-address': 'fullStreetAddress',
  tel: 'telephoneNumber',
  url: 'URL',
  username: 'username'
};
```

Example:

```jsx
<TextArea
  autoComplete="email"
  inputMode="email"
/>

<TextArea
  multiline
  readOnly
  rows={3}
/>
```

> ### New styles

Author: Add desire for CSS animation and transition support too

* CSS Lengths
  * Support `em`, `rem`, `v*`, `px` length units.
  * This can be shimmed in user-space on top of the existing `StyleSheet` API, with a React Context used to provide ancestral font-size information for `rem` values.
* CSS Colors
  * Support CSS Colors using [Colorjs.io](https://colorjs.io/).
  * This can be shimmed in user-space.
* CSS Media Queries
  * Support CSS Media Queries. Although Media Queries are not a preferred long-term solution, [Container Queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Container_Queries) are not yet widely supported by browsers.
  * This can be shimmed in user-space on top of the existing `StyleSheet` API.
* CSS Logical Properties
  * Support CSS logical property names, i.e., `paddingInlineStart` rather than only non-standard `paddingStart`.
  * Support subtree-level logical properties, where setting the `dir` prop to `ltr` or `rtl` alters the way logical properties are resolved. Currently, React Native only supports global writing direction configuration for Android.
* `aspectRatio`
  * Support string values, i.e., `'16 / 9'`, to align with CSS.
* `backgroundImage`
  * Support setting background images via `url()`.

>   * Optional: support CSS gradients.

Reviewer 2: Similar to the [blur suggestion](https://github.com/react-native-community/discussions-and-proposals/pull/496/files#r944749646), I think there's value in adding basic gradient support while still having value for community packages to add features like `startPosition` and `endPosition` which are more powerful than CSS's `angle`.

* `boxShadow`
  * Add native support for CSS box shadows.

>   * This would replace the very buggy iOS-specific `shadow*` styles and Android `elevation` style, allowing developers to unify shadows across native and web platforms.

Reviewer 2: - `inset` shadows would also be quite useful.
- Multiple shadows are supported in SwiftUI, meaning it's substantially easier to reimplement `box-shadow` in SwiftUI than it is in React Native.

* `fontVariant`
  * Support space-separated string values to align with CSS.
* `fontWeight`

>   * Support number values to align with React DOM / CSS.

Reviewer 2: ```suggestion
  * Support number values to align with React DOM / CSS.
  * Work reliably on Android.
```

* `lineClamp`
  * Support CSS clamping text to a set number of lines via style.
  * This would be equivalent to using the `numberOfLines` prop on `<Text>` components.
* `objectFit`
  * Support CSS `objectFit` property for `<Image>`.
  * This would be a subset (no "repeat" keyword) of functionality currently supported by the `resizeMode` prop and style for `<Image>`.
  * `objectFit === 'contain'` => `resizeMode = 'contain'`
  * `objectFit === 'cover'` => `resizeMode = 'cover'`
  * `objectFit === 'fill'` => `resizeMode = 'stretch'`
  * `objectFit === 'scale-down'` => `resizeMode = 'contain'`
* `pointerEvents`
  * Support setting the `pointerEvents` values via style rather than a prop.
  * Retain the React Native specific `box-none` and `box-only` values.
* `textShadow`
  * Add native support for CSS text shadows.
* `transform`
  * Support using string values to set transforms.
  * This can be shimmed in user-space with a parser.

> * `verticalAlign`

Reviewer 2: I wouldn't be heartbroken if `verticalAlign` is deprioritized/never added. It's up there next to `float` in terms of unintuitive CSS APIs.

  * Make `verticalAlign` the same as `textAlignVertical`.
  * `verticalAlign === 'middle'` is the same as `textAlignVertical = 'center'`;
* `userSelect`
  * Support setting ability to select text via style.

>   * Equivalent to using `selectable` prop on `<Text>`.

Reviewer 2: Would love to see [`position`](https://developer.mozilla.org/en-US/docs/Web/CSS/position) fully implemented, specifically `fixed` and `sticky` are annoying to swap out when doing web/native.

> 

Reviewer 2: Some basic `filter: blur()` support would also be useful now that Android 12 has native support. It's pretty frustrating to have to install a new native package and rebuild to obtain this functionality, especially when it's built into the OS.

Standalone blur libraries would still have value by providing more platform-specific blur functionality, but the basic web parity would feel really nice.

```jsx
<View
  style={{
    boxShadow: '1px 1px 1px 1px #eee',
    backgroundColor: 'hsla(50,50,50,0.5)',
    pointerEvents: 'none',
    transform: 'scale(0.9)',
    width: '5rem'
  }}
>

<Text
  style={{
    lineClamp: 3,
    userSelect: 'none',
    verticalAlign: 'middle'
  }}
>

<Image
  style={{
    objectFit: 'cover'
  }}
>
```

### New imperative APIs

* `EventTarget` API

>   * The `EventTarget` API is a implemented in JS environments, and should be available on React Native host elements. This feature is commonly used on web, and could be used by React Native developers to support more complex features without first requiring further core API changes to support each use case.

Reviewer 8: ```suggestion
  * The `EventTarget` API is implemented in JS environments, and should be available on React Native host elements. This feature is commonly used on web, and could be used by React Native developers to support more complex features without first requiring further core API changes to support each use case.
```

  * Promote `node.unstable_addEventListener`, `node.unstable_removeEventListener` to stable APIs.
  * Add `node.dispatchEvent` support.
  * This cannot be shimmed in user-space.
* `node.getBoundingClientRect`
  * Synchronous element measurement API. Current APIs like `node.measure` are async and return slightly different information to what you expect from `getBoundingClientRect` on web.
  * This cannot be shimmed in user-space.

### Other

* Revisit the type of all events modeled on W3C events. For example, `onTouchStart` should receive an event object that conforms to that received by the same prop in React DOM.

> * Aim to deprecate `StyleSheet.flatten()`.

Reviewer 2: As long as there are missing CSS properties this will have value. I use it in project's to introspect and replace `position: 'fixed'` with `position: 'absolute'` on native platforms. Seems like a moving target that's potentially hard to keep up with.

Author: That introspection should not be done at runtime, as it is slow and web will not support it.

  * This API encourages runtime introspection of styles in render calls. This pattern is a performance cost (flattening arrays of objects), and prevents us from supporting build-time optimizations for web (extracting styles to CSS files).
* Aim to deprecate `StyleSheet.compose()`.
  * This API has no real value.
* Aim to restore obfuscation of styles passed to `StyleSheet.create()`.
  * Returning the styles from create calls encourages introspection of those styles. On web, we need to be able to remove the JavaScript style objects from bundles to support build-time optimization like extraction to CSS files.
* Aim to deprecate the `Touchable*` components.
  * The `Pressable` component was shipped as a replacement for `Touchable*`.
* Aim to support more interaction states with `Pressable`.
  * Add ‘hovered’ and ‘focused’ to the state object, and replace ‘testOnly_pressed’ with an object of states.
* Aim to move `Animated` out of core.

>   * Move Animated out of core to separate package in monorepo, allow React Native for Web to import this package. Eventually could drop Animated from ‘react-native’ export.

Reviewer 2: I got no feedback when [trying this two years ago](https://github.com/facebook/react-native/pull/28892). Still agree that this is a good move and also agree it should stay in the facebook/react-native repo.

* Aim to move `VirtualizedList` and `FlatList` out of core.
  * Move `VirtualizedList`/`FlatList` out of core to separate package in monorepo, allow React Native for Web to import this package. Eventually could drop these components from `'react-native'` export.

> * Compile to standard JavaScript before publishing to npm (make this the default for the community). React Native and its community packages currently publish uncompiled, flow-typed JavaScript to npm. This causes common tools to crash on these packages.

Reviewer 2: I'd love to understand why this wasn't done years ago. Metro would run a lot faster if it didn't have to transpile everything all the time. 

This one feature might warrant its own proposal given the scope. Perhaps @motiz88 also has some thoughts on this.

Reviewer 9: Yes, I second making this its own proposal - the devil's in the details here.

1. If the entire ecosystem switched to publishing standard JavaScript to npm overnight (which realistically would be a complicated migration in its own right) - and if that in practice means compiling to the lowest common denominator - there's a real risk of bloating bundle sizes and losing out on runtime performance because of poorly/inconsistently transpiled packages. (Think packages bringing their own incompatible versions of Babel helpers, or transpiling features that the target runtime actually supports natively.) Where now we have leverage to compile things in the best way for each app, we would spread that leverage across a long tail of community-maintained packages. So IMO we'd need to give clear guidance (+ recommend specific tools/configs?) to package authors re: what and how much to transpile, so we can still influence the overall outcome for app authors.
2. By "standard JavaScript" are we also thinking of moving away from React Native-isms like platform-specific file extensions, asset scale file extensions (`@2x.png` etc), and reliance on spec-breaking compiler features such as [inline requires](https://reactnative.dev/docs/ram-bundles-inline-requires#inline-requires) and `Platform.select()` inlining? IMO these are good to explore in principle, but my main point is that we should set a clear goal re what "standard JS" means and to what exact extent "common tools" (which?) should be able to process these packages.

Author: I don't think it needs to be that complicated. We should align with what libraries targeting web or authored in typescript are already doing. RN JS crashes when most tools try to parse it because it contains flow types in JS files. TS packages on npm don't have this problem because compiling `*.ts` to `*.js` + `*.d.ts` is part of the workflow for TS projects. So let's start by getting flow types out of the published JS files.

Reviewer 9: > let's start by getting flow types out of the published JS files.

To be clear, I'm fully on board with this part. It's just that there's more related scope that's worth being clear about. (e.g. just removing Flow types doesn't mean we stop transpiling or speed up builds)

Reviewer 9: Concretely, I would recommend that we narrow the wording here:

```suggestion
* Remove Flow types before publishing to npm (make this the default for the community). React Native and its community packages currently publish uncompiled, flow-typed JavaScript to npm. This causes common tools to crash on these packages.
```

Reviewer 10: Many packages already use `react-native-builder-bob` for compiling code before publishing. So if there's a recommended configuration on how to compile, it'd be a good place to do so and will benefit a lot of libraries with a relatively simple way to migrate.

https://github.com/callstack/react-native-builder-bob

* ARIA props
  * Support `aria-*` aliases for `accessibility*` props.
  * `aria-busy` => `accessibilityState.busy`
  * `aria-checked` => `accessibilityState.checked`
  * `aria-disabled` => `accessibilityState.disabled`
  * `aria-expanded` => `accessibilityState.expanded`
  * `aria-hidden` => `accessibilityElementsHidden` (iOS) + `importantforAccessibility` (Android)
  * `aria-label` => `accessibilityLabel`
  * `aria-live` => `accessibilityLiveRegion` (and add `'off'` value => `'none'`)
  * `aria-modal` => `accessibilityViewIsModal`
  * `aria-selected` => `accessibilityState.selected`
  * `aria-valuemax` => `accessibilityValue.max`
  * `aria-valuemin` => `accessibilityValue.min`
  * `aria-valuenow` => `accessibilityValue.now`
  * `aria-valuetext` => `accessibilityValue.text`
  * `role` => `accessibilityRole`
  * See prior work: [RFC Accessibility APIs](https://github.com/react-native-community/discussions-and-proposals/pull/410/files).
* `id`

>   * Add `id` prop as synonym for `nativeID`.

Reviewer 11: is the use of the word 'synonym' instead of 'alias' here (and in several other places in this rfc) intentional? if so, can you clarify the difference in meaning?

  * This would be equivalent to using the `numberOfLines` prop on `<Text>` components.
* `objectFit`
  * Support CSS `objectFit` property for `<Image>`.
  * This would be a subset (no "repeat" keyword) of functionality currently supported by the `resizeMode` prop and style for `<Image>`.
  * `objectFit === 'contain'` => `resizeMode = 'contain'`
  * `objectFit === 'cover'` => `resizeMode = 'cover'`
  * `objectFit === 'fill'` => `resizeMode = 'stretch'`
  * `objectFit === 'scale-down'` => `resizeMode = 'contain'`
* `pointerEvents`
  * Support setting the `pointerEvents` values via style rather than a prop.
  * Retain the React Native specific `box-none` and `box-only` values.
* `textShadow`
  * Add native support for CSS text shadows.
* `transform`
  * Support using string values to set transforms.
  * This can be shimmed in user-space with a parser.
* `verticalAlign`
  * Make `verticalAlign` the same as `textAlignVertical`.
  * `verticalAlign === 'middle'` is the same as `textAlignVertical = 'center'`;
* `userSelect`
  * Support setting ability to select text via style.
  * Equivalent to using `selectable` prop on `<Text>`.

> 

Reviewer 2: `borderRadius` is also missing support for strings `1px 2px 3px` [ref](https://developer.mozilla.org/en-US/docs/Web/CSS/border-radius).

Author: Multi value shorthands are not supported in React Native

> 

Reviewer 2: Seems simple, but just for the sake of completeness we should also add `visibility` which should be easy to add.

```jsx
<View
  style={{

>     boxShadow: '1px 1px 1px 1px #eee',

Reviewer 12: Wouldn't it be strange to have _just this one_ CSS-like shorthand (i.e. multiple values separated by spaces) in RN? I understand where this comes from (there's no long-form `box-shadow` in CSS as far as I know), but it seems out of space.

I'd rather this read `boxShadow: [1, 1, 1, 1, 'red']`, and then extend this syntax to at least some common places where in CSS you'd use a shorthand, for example:

```js
margin: [4, 10],
padding: [8, '20%', 15],
border: [1, 'solid', 'red'],
```

having shorthands for these common properties would really improve RN StyleSheet ergonomics, and map cleanly to CSS (while not pretending to be CSS)

Author: The point is to avoid introducing non-standard syntax that requires transforms on web and reeducation of developers

  * This API encourages runtime introspection of styles in render calls. This pattern is a performance cost (flattening arrays of objects), and prevents us from supporting build-time optimizations for web (extracting styles to CSS files).
* Aim to deprecate `StyleSheet.compose()`.
  * This API has no real value.
* Aim to restore obfuscation of styles passed to `StyleSheet.create()`.
  * Returning the styles from create calls encourages introspection of those styles. On web, we need to be able to remove the JavaScript style objects from bundles to support build-time optimization like extraction to CSS files.
* Aim to deprecate the `Touchable*` components.
  * The `Pressable` component was shipped as a replacement for `Touchable*`.
* Aim to support more interaction states with `Pressable`.
  * Add ‘hovered’ and ‘focused’ to the state object, and replace ‘testOnly_pressed’ with an object of states.
* Aim to move `Animated` out of core.
  * Move Animated out of core to separate package in monorepo, allow React Native for Web to import this package. Eventually could drop Animated from ‘react-native’ export.
* Aim to move `VirtualizedList` and `FlatList` out of core.

>   * Move VirtualizedList/FlatList out of core to separate package in monorepo, allow React Native for Web to import this package. Eventually could drop Animated from ‘react-native’ export.

Reviewer 13: > Eventually could drop Animated from ‘react-native’ export.

I guess here it should be drop `VirtualizedList` and `FlatList` from react-native export? not sure

* Remove Flow types and JSX before publishing to npm (make this the default for the community). React Native and its community packages currently publish uncompiled, flow-typed JavaScript to npm. This causes common tools to crash on these packages.

> * Prevent libraries importing `'react-native'` package internals, by using the Node.js package exports API.

Reviewer 12: Please don't. While doing this is a practice that should be avoided as much as possible, there is real value in being able to quickly patch up or work around an issue, and that often involves digging around the internals. Making `react-native` more opaque might make feel us better about its cleanliness, but it will just make working around issues take more work

Author: You can use things like patch-package to do your work arounds. But libraries should not be depending on unstable internal APIs. If there are specific asks for new public APIs, open issues with the use case

- [ ] `onPointerLeave`.
- [ ] `onPointerMove`.
- [ ] `onPointerOut`.
- [ ] `onPointerOver`.
- [ ] `onPointerUp`.
- [ ] `onKeyDown`.
- [ ] `onKeyUp`.

Example:

```jsx
<View
  aria-hidden={false}
  aria-label="Accessibility label"
  id="native-id"
  on-click={handleDOMClick}
  on-pointerdown={handlePointerDown}
  role="button"
  tabIndex={0}
>
```

## `Image` component

Inherits all props and events from "Core Component".

### Primitive props

Various React DOM `img` props have direct equivalents in React Native. For example, `srcSet` is a subset of the `source` prop but without support for specifying headers, etc. Features like alternative text content are also required to target web.

- [x] `alt` prop for alternative text support.
- [x] `crossOrigin` is equivalent to setting the relevant `Header` in the `source` object. Declaratively configure cross-origin permissions for loaded images.
- [ ] `decoding`.
- [ ] `fetchPriority`.
- [x] `height` is the same as `source.height`.
- [ ] `loading`.
- [x] `referrerPolicy` is equivalent to setting the relevant `Header` in the `source` object. Declaratively configure referrer policy.
- [x] `src` is the same as `source.uri`.

> - [x] `srcSet` is the same as setting a `source` array with `uri` and `scale` properties defined.

Reviewer 14: Should RN support `sizes`? https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/sizes

This will allow loading different images based upon device dimensions, a very useful feature when developing for both mobile and desktop

Author: Possibly. At Meta we have no examples of that kind of feature being used on either web or native. So for now I think it's more a feature we should attempt to shim in user-space outside of React Native.

This proposal aims to allow developers to target more platforms with cross-platform APIs, and deliver better performance when targeting browsers. Features for Android, iOS, and Web are unified by aligning with the Web standard. Support standards helps to:

* minimize the overhead when running in browsers;
* reduce developer education required to learn features;
* set clear and cohesive API end-state expectations for contributors to aim for;
* support backwards compatibility;
* develop universal codebases.

React Native for Web uses React DOM and native DOM APIs to shim React Native components and APIs on the web. The tooling for an existing React DOM app simply maps the 'react-native' export to 'react-native-web'. This is the most complete and widely used shim, but comes with considerable DX and UX costs on the web.

![](https://user-images.githubusercontent.com/239676/205388174-6fce19f1-8b3a-4c6e-82f1-89ef9c21ab7d.png)

Implementing a Strict DOM subset in React Native shifts the weight of bridging native and web apps onto React Native Native, where it can be done most efficiently. On the web, we would combine React DOM with stylex.

> ![](https://user-images.githubusercontent.com/239676/205388313-e58f8793-51cd-4a6e-86e7-89425fd2d0e1.png)

Reviewer 14: I think the `stylex` apis need to be refined first, especially with media/container queries & animations which are still in RFC, have had minimal community input and have issues with theming.

- [ ] [`hidden`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/hidden).
- [x] [`id`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id) is the same as `nativeID`.
- [ ] [`inert`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inert).
- [x] [`role`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles) is the same as `accessibilityRole` (with support for web values).
- [x] `style` is unchanged.
- [x] [`tabIndex`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex) is the same as `focusable` (Android, Desktop). Add support for `0` and `-1` values only.

### Event props

Props for commonly handled events should also be supported on core components. Each event handler should receive a W3C event object of the corresponding type, not a synthetic event. These props are merely conveniences and otherwise identical to using the `EventTarget` API mentioned later on. Note that no capture phase props are included, as listening to the capture phase (as well as custom events) should be handled with the `EventTarget` API.

> To avoid breaking changes in React Native and React DOM, it may be preferrable to create new event prop names that receive events as specified by web standards, e.g., `on-click`, `on-pointerup`.

Reviewer: Can we flush this out a bit more?  How complete are we looking at making `addEventListener`?  In order to support `stopPropagation` and `preventDefault` the events would need to be on the UI thread.  Thats possible in fabric, but likely have some pretty serious ramifications.

And then is the idea that we'd have old onPointerUp that works in the current RN manor, and a new on-pointerup which would happen on the UI thread and allow all the `stopPropagation` and `preventDefault` functionality of W3C events.

Would we support custom events too?  Would those work on the UI thread also, so that they can potentially interact with native components?

  if (crossOrigin === 'use-credentials') {
    headers['Access-Control-Allow-Credentials'] = true;
  }
  if (referrerPolicy != null) {
    headers['Referrer-Policy'] = referrerPolicy;
  }
  source.push({ headers, height, scale, uri, width });
});
```

-----

# Styles API (CSS subset)

The styling capabilities of React Native should be significantly expanded to cover more of the features that are heavily relied upon by web engineers. Furthermore, React Native needs to gracefully handle unknown style properties or values as is expected on the web. Rather than hard crashing it could print warnings to bring attention to any styles that are not currently supported on native.

## StyleSheet

* [ ] Deprecate `StyleSheet.absoluteFill`. Use `position:'absolute';inset:0;` instead.
* [ ] Deprecate `StyleSheet.absoluteFillObject`.
* [ ] Deprecate `StyleSheet.compose()`. Use existing `[ a, b ]` syntax instead.
* [ ] Deprecate `StyleSheet.flatten()`. This API encourages runtime introspection of styles in render calls. This pattern is a performance cost (flattening arrays of objects), and prevents us from supporting build-time optimizations for web (e.g., extracting styles to CSS files).

> * [ ] Deprecate or rename `setStyleAttributePreprocessor` (e.g., `unstable_setStyleAttributePreprocessor`).

Reviewer 2: Expo uses this to [sandbox the `fontFamily` property](https://github.com/expo/expo/blob/f69f681116a94ea85b8a5d9633f6b54e3595f3e1/packages/expo/src/Expo.fx.tsx#L14-L17).

* [ ] `float`.
* [ ] [`justifySelf`](https://developer.mozilla.org/en-US/docs/Web/CSS/justify-self). Sets the way a box is justified inside its alignment container along the appropriate axis.
* [ ] (`lineClamp` is equivalent to `numberOfLines` prop on `<Text>` components).
* [ ] [`objectPosition`](https://developer.mozilla.org/en-US/docs/Web/CSS/object-position).
* [ ] `overflowBlock`.
* [ ] `overflowInline`.
* [ ] `overflowX`.
* [ ] `overflowY`.
* [ ] [`placeContent`](https://developer.mozilla.org/en-US/docs/Web/CSS/place-content). Shorthand for setting `alignContent` and `justifyContent`. Value is an enum of `baseline`, `first baseline`, `last baseline`, `center`, `end`, `flex-start`, `flex-end`, `left`, `right`, `safe`, `space-around`, `space-between`, `space-evenly`, `start`, `stretch`, `unsafe`. Note that `space-evenly` should also be supported for `alignContent` and `justifyContent`.
* [ ] [`placeItems`](https://developer.mozilla.org/en-US/docs/Web/CSS/place-items). Shorthand for setting `alignItems` and `justifyItems`. Value is an enum of `auto`, `baseline`, `first baseline`, `last baseline`, `center`, `end`, `flex-start`, `flex-end`, `left`, `right`, `start`, `stretch`.

> * [ ] [`placeSelf`](https://developer.mozilla.org/en-US/docs/Web/CSS/place-self). Shorthand for setting `alignSelf` and `justifySelf`. Value is an enum of `auto`, `baseline`, `first baseline`, `last baseline`, `center`, `end`, `flex-start`, `flex-end`, `left`, `normal`, `right`, `start`, `stretch`.

Author: These might be relatively easy to implement (other than the full list of values) as they expand into properties with existing support.

* [ ] [`rotate`](https://developer.mozilla.org/en-US/docs/Web/CSS/rotate).

> * [ ] [`scale`](https://developer.mozilla.org/en-US/docs/Web/CSS/scale).

Author: The transform properties might be shimm-able.

* [ ] `textShadow`. Add native support for CSS text shadows.
* [ ] [`touchAction`](https://developer.mozilla.org/en-US/docs/Web/CSS/touch-action).
* [ ] [`translate`](https://developer.mozilla.org/en-US/docs/Web/CSS/translate).
* [ ] `visibility`. Add support for visually hiding elements.
* [ ] [`whiteSpace`](https://developer.mozilla.org/en-US/docs/Web/CSS/white-space).

> ## CSS Transitions

Author: This could be a win for DX and UX. SImpler transitions than using `Animated`, less JS in bundles, and potential improvements to optimization on the native side.

Support declarative transitions that can be optimized on the native side and avoid the need for `Animated`. Support `transitionProperty`, `transitionDuration`, `transitionTimingFunction`, and `transitionDelay` styles (see [transition](https://developer.mozilla.org/en-US/docs/Web/CSS/transition)). Consider dispatching the corresponding W3C transition events too.

-----

# Environment API

## document

Support the `document` object for common patterns such as listening to document resize, listening to capture or bubble phase events for the entire application (e.g., "outside click" pattern), etc.

### Properties

- [ ] [`activeElement`](https://developer.mozilla.org/en-US/docs/Web/API/Document/activeElement).
- [ ] [`visibilityState`](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState)

### Events

> * [ ] `scroll`.

Author: `fixed` and `sticky` can be relative to non-root nodes. [expo asked](https://github.com/react-native-community/discussions-and-proposals/pull/496#discussion_r944746656) for both these values, and I think `sticky` in particular can be useful for implementing sticky headers.

the wider question of what to do about default root scroll view is something else we could also figure out.

* [ ] `visibilitychange`.

## IntersectionObserver

Support the [`IntersectionObserver`](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver) API for observing the intersection of target elements. This can also be used as a building block for performance tooling.

## MutationObserver

Support the latest [`MutationObserver`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) API for watching changes to the host node tree. This can also be used as a building block for performance tooling.

> ## ResizeObserver

Author: This is related to Container Queries. Implementing `ResizeObserver` should help to address some of the differences between `onLayout` data on web and native: whether the effect of transforms are included in dimensions, and how coordinates related to scroll position. `ResizeObserver` is only concerned with width/height and could back a new `onResize` prop.

Further below, the proposal includes `getBoundingClientRect`, which is another opportunity to unify the platforms. React Native currently only supports async measure methods on elements. If a developer needs the coordinates of an element at the time of its resize, they can call `getBoundingClientRect` in the resize callback.

-----

# Components API (DOM subset)

A logical extension of support more React DOM props is to support a subset of React DOM components directly.

This would provide a built-in declarative API for cross-platform elements in React Native. All unknown tags default to `<span>` layout. All known-but-unimplemented block-level tags default to `<div>` layout. Most tags can be mapped to existing React Native components and props, in the reverse of the [mapping used by React Native for Web](https://github.com/necolas/react-native-web/blob/b2eb3ca03ffc4b8570591b6c2082c6e4edae0e7f/packages/react-native-web/src/modules/AccessibilityUtil/propsToAccessibilityComponent.js#L12-L30).

## Sections

Content sectioning elements allow you to organize the document content into logical pieces. Use the sectioning elements to create a broad outline for your page content, including header and footer navigation, and heading elements to identify sections of content.

* [ ] `article` is equivalent to `View` with `display:block` layout and `role="article"`.
* [ ] `aside`
* [ ] `div` is equivalent to `View` with `display:block` layout.
* [ ] `footer`.
* [ ] `h1-6`.
* [ ] `header`.
* [ ] `main`.
* [ ] `nav`.

> * [ ] `section` is equivalent to `View` with `display:block` layout and `role="section"`.

Author: These elements are trivial to shim in React Native, other than the `block` layout.

Author: In the long term, I'm not sure. I think Seb suggested they could be "just there" like they are in react-dom. But in the immediate term, we might want/need to import them explicitly and in a way that can be recognized as a component (e.g., `<strict.div>` or `<html.div>`), because we won't be able to make changes to `style` or event props in the react-dom components. This would mean that this API is also opt-in for web users.

## Text

Text content elements organize blocks or sections of content. Important for accessibility, these elements identify the purpose or structure of that content.

* [ ] `blockquote`.
* [ ] `li`.
* [ ] `ol`.

> * [ ] `ul`.

Author: RN should probably render these with no styles by default. On the web side, we'll reset the styles as is typical for web app development.

Reviewer 15: Interesting… I had never considered the necessity for web compliance to also involve either implementing default styles or including a reset stylesheet. (It's been so long since I've had to set one of those up…)

I *think* I agree with assuming common reset practices…

* [ ] `p`.
* [ ] `pre`.

## Inline text

Inline text semantics define the meaning, structure, or style of a word, line, or any arbitrary piece of text.

> * [ ] `a`.

Author: Possibly the most complicated element to implement?

Reviewer 15: Yeah… I'd almost include an annotation here to acknowledge that `a` should probably not be supported (or at least, not any `a`-specific attributes like `href`) until we have a clear proposal for how to deal with navigation (and maybe routing).

## Forms

Elements which can be used together to create forms which the user can fill out and submit. There's a great deal of further information about this available in the [HTML forms guide](https://developer.mozilla.org/en-US/docs/Learn/Forms).

* [ ] `button`.
* [ ] `input` is equivalent to `TextInput` with `inputMode` corresponding to `type`.
* [ ] `label`.
* [ ] `optgroup`.
* [ ] `option`.
* [ ] `progress`.
* [ ] `select`.
* [ ] `textarea` is equivalent to `TextInput` with `multiline={true}`.

-----

# Elements API (DOM subset)

Subset of cross-platform DOM APIs exposed on element instances, document, etc. This would be an imperative API for working with the Shadow tree in React Native. Based on data from [MDN Web API](https://developer.mozilla.org/en-US/docs/Web/API), please refer to the MDN links for complete DOM APIs and comment if you believe any should be included in the proposed subset.

In practice, this DOM API subset is generally limited to the “best practice” APIs that can safely be used with React DOM (i.e., read-only operations that don’t modify the DOM tree), and most React components should already be limited to this subset.

> ## EventTarget

Author: This API is common to make JS environments, and in the context of React Native it provides a lot of developer flexibility for building event handling and dispatching systems. A lot of React hooks built for React DOM could "just work" with the EventTarget API alone.

The `EventTarget` API is implemented in JS environments, and should be available on React Native host elements. This feature is commonly used on web, and could be used by React Native developers to support more complex features without first requiring further core API changes to support each use case.

### Methods

- [ ] `EventTarget.addEventListener()`. Registers an event handler to a specific event type on the element.
- [ ] `EventTarget.dispatchEvent()`. Dispatches an event to this node in the DOM and returns a boolean value that indicates whether no handler canceled the event.

> - [ ] `EventTarget.removeEventListener()`. Removes an event listener from the element.

Author: We already have an `unstable_*` implementation of the listener methods in React Native. We'd need to promote them to stable and implement the missing `dispatchEvent` function along with `new Event()` and `new CustomEvent()`

- [ ] `Element.scrollLeftMax` [Read only]. Returns a number representing the maximum left scroll offset possible for the element.
- [ ] `Element.scrollTop`. A number representing number of pixels the top of the element is scrolled vertically.
- [ ] `Element.scrollTopMax` [Read only]. Returns a number representing the maximum top scroll offset possible for the element.
- [ ] `Element.scrollWidth` [Read only]. Returns a number representing the scroll view width of the element.

### Methods

Element inherits methods from its parents Node, and its own parent, EventTarget.

- [ ] `Element.computedStyleMap()`. Returns a StylePropertyMapReadOnly interface which provides a read-only representation of a CSS declaration block that is an alternative to CSSStyleDeclaration.
- [ ] `Element.getBoundingClientRect()` **P1**. Returns the size of an element and its position relative to the viewport.
- [ ] `Element.scroll()`. Scrolls to a particular set of coordinates inside a given element.
- [ ] `Element.scrollBy()`. Scrolls an element by the given amount.
- [ ] `Element.scrollIntoView()`. Scrolls the page until the element gets into the view.

> - [ ] `Element.scrollTo()`. Scrolls to a particular set of coordinates inside a given element.

Author: These properties are often used in event handling.

Reviewer 15: There have been [some early internal discussions](https://fburl.com/d1pak0i8) about a platform-agnostic React Component Instance API that makes the "best practice" imperative APIs available while abstracting away most of the more intrusive tree traversal APIs (which violation encapsulation of React components).

A lot of the APIs you've identified here as typically valuable in event handling, were also included in that.

Author: Need to add:

* `getClientRects()`
* `compareDocumentPosition()`

Will keep `contains()` and `getBoundingClientRect()` because they are used far more frequently than the above, and easy to derive. And will not add the non-standard APIs in that proposal.

- [ ] `HTMLElement.hidden` [Read only]. A boolean value indicating if the element is hidden or not.
- [ ] `HTMLElement.offsetHeight` [Read only]. Returns a double containing the height of an element, relative to the layout.
- [ ] `HTMLElement.offsetLeft` [Read only]. Returns a double, the distance from this element's left border to its offsetParent's left border.
- [ ] `HTMLElement.offsetParent` [Read only]. Returns a Element that is the element from which all offset calculations are currently computed.
- [ ] `HTMLElement.offsetTop` [Read only]. Returns a double, the distance from this element's top border to its offsetParent's top border.
- [ ] `HTMLElement.offsetWidth` [Read only]. Returns a double containing the width of an element, relative to the layout.

### Methods

Inherits methods from its parent, Element, and implements those from DocumentAndElementEventHandlers, GlobalEventHandlers, and TouchEventHandlers.

- [x] `HTMLElement.blur()`. Removes keyboard focus from the currently focused element.
- [ ] `HTMLElement.click()`. Sends a mouse click event to the element.

> - [ ] `HTMLElement.focus(options)`. Makes the element the current keyboard focus.

Author: `focus` options are currently only `preventScroll` and `focusVisible`.

- [ ] `lostpointercapture`. Fired when a captured pointer is released. Also available via the `onLostPointerCapture` prop.
- [ ] `pointercancel`. Fired when a pointer event is canceled. Also available via the onpointercancel property.
- [ ] `pointerdown`. Fired when a pointer becomes active. Also available via the onpointerdown property.
- [ ] `pointerenter`. Fired when a pointer is moved into the hit test boundaries of an element or one of its descendants. Also available via the onpointerenter property.
- [ ] `pointerleave`. Fired when a pointer is moved out of the hit test boundaries of an element. Also available via the onpointerleave property.
- [ ] `pointermove`. Fired when a pointer changes coordinates. Also available via the onpointermove property.
- [ ] `pointerout`. Fired when a pointer is moved out of the hit test boundaries of an element (among other reasons). Also available via the `onPointerOut` prop.
- [ ] `pointerover`. Fired when a pointer is moved into an element's hit test boundaries. Also available via the `onPointerOver` prop.

> - [ ] `pointerup`. Fired when a pointer is no longer active. Also available via the `onPointerUp` prop.

Author: All these events (not just pointer) are events that must be W3C compatible from the beginning. Nothing that goes through `EventTarget` should be using the non-standard types for events, even if the `on{Event}` prop API keeps using the non-standard types.

- [ ] `selectionchange`. Fires when the text selection in a `<textarea>` element has been changed.  Also available via the `onSelectionChange` prop.

-----

# Other maintenance

Other tasks that will help to reduce API fragmentation within React Native, and between React Native and React Native for Web. Large, user-space modules currently exported from the `react-native` package could eventually be moved entirely to separate packages, reducing the size of the core library on native and web, and allowing alternative solutions to compete on a level playing field. JavaScript from the React Native ecosystem should not be published to npm without removal of non-standard syntax.

* [ ] Deprecate the `Touchable*` components. Use `Pressable`.
* [ ] Add `hovered` state support to `Pressable`.
* [ ] Add `focused` state support to `Pressable`.
* [ ] Move `Animated` out of core and into a separate package in the monorepo.
* [ ] Move `VirtualizedList` out of core and into a separate package in the monorepo.
* [ ] Move `VirtualizedSectionList` out of core and into a separate package in the monorepo.
* [ ] Move `FlatList` out of core and into a separate package in the monorepo.

> * [ ] Move `SectionList` out of core and into a separate package in the monorepo.

Author: These tasks are in progress

};
```

`srcSet` to `source` mapping:

```js
const { crossOrigin, referrerPolicy, srcSet } = props;
const source = [];
const srcList = srcSet.split(', ');
srcList.forEach((src) => {
  const [uri, xscale] = src.split(' ');
  const scale = parseInt(xscale.split('x')[0], 10);
  const headers = {};
  if (crossOrigin === 'use-credentials') {
    headers['Access-Control-Allow-Credentials'] = true;
  }
  if (referrerPolicy != null) {
    headers['Referrer-Policy'] = referrerPolicy;
  }
  source.push({ headers, height, scale, uri, width });
});
```

-----

# Styles API (CSS subset)

## StyleSheet

* [ ] Deprecate `StyleSheet.absoluteFill`. Use `position:'absolute';inset:0;` instead.
* [ ] Deprecate `StyleSheet.absoluteFillObject`.

> * [ ] Deprecate `StyleSheet.compose()`. Use existing `[ a, b ]` syntax instead.

Reviewer 16: Compose is useful for internal micro optimizations and avoid array allocations. Maybe we could make it an internal-only API, which is stubbed / ignored for web?

Author: How much of a difference do those micro optimisations make it product code rarely uses it? Not to mention the API is limited to only 2 args

> * [ ] Deprecate `StyleSheet.flatten()`. This API encourages runtime introspection of styles in render calls. This pattern is a performance cost (flattening arrays of objects), and prevents us from supporting build-time optimizations for web (e.g., extracting styles to CSS files).

Reviewer 16: Could be made internal-only too. Flattening is required as part of the native interface / React's prop diffing.

Author: Yeah, as long as it's not being used in user space

Reviewer 10: When working on a library, it's useful to be able to accept a regular-style prop and flatten it to pass it down to the underlying native component. If this is removed, then we'd need to either flatten manually or force the user to pass a plain object for `style` - which isn't ideal since the different kind of API won't be intuitive, or user would need to manually spread the objects.

Reviewer 10: On web we don't need the individual style properties as passing the class name to the tag takes care of applying multiple styles.

With a native component (e.g. if I implement my own `Image` component), we want a regular object that we use to apply the styles on the native side. It's the same reason why React Native core does it for the built-in components.

Author: That's something that could be provided without having JS `flatten` exist in user space and being used for other purposes that cause performance issues.

Reviewer 10: Yeah, just something that needs to exist before removing this API

Author: Deprecation != removal. And flattening is very straight-forward, so you could definitely do this / copy-paste in user-space too.

* [ ] Deprecate or rename `setStyleAttributePreprocessor` (e.g., `unstable_setStyleAttributePreprocessor`).
* [ ] `StyleSheet.create()` should obfuscate styles to prevent introspection (i.e., revert identify function change). On web, we need to be able to remove the JavaScript style objects from bundles to support build-time optimization like extraction to CSS files.

## CSS compatibility

* [x] `aspectRatio`. Support string values, i.e., `'16 / 9'`, to align with CSS.

> * [ ] `backgroundImage`. Add support for setting background images via `url()`. Stretch: support CSS gradients.

Author: Perhaps it could switch out for ImageBackground based on styles. This is a basic web feature after all
